File: QuickSort/quick_sort.py
from typing import List

class QuickSort:
    """ Quick sort algorithm implementation. """

    @staticmethod
    def _partition(a: List[int], low: int, high: int):
        """ Partition function. 
        
        Parameters:
        -----------
        a : List[int]
            List to be sorted.
        
        low : int
            Lower list bound.
        
        high : int
            Higher list bound.

        Returns: int
            Pivot of sorting algorithm.

        """
        pivot = a[high]
        i = low - 1

        for j in range(high):
            if a[j] <= pivot:
                i += 1
                a[i], a[j] = a[j], a[i]
        
        a[i + 1], a[high] = a[high], a[i + 1]
        return i + 1
    
    @staticmethod
    def sort(a: List[int], low: int, high: int):
        """ Sorting function. 
        
        Parameters:
        -----------
        a : List[int]
            List to be sorted.
        
        low : int
            Lower list bound.
        
        high : int
            Higher list bound.

        """
        if low < high:
            partition_idx = QuickSort._partition(a, low, high)

            # Sort first half
            QuickSort.sort(a, low, partition_idx - 1)

            # Sort second hald
            QuickSort.sort(a, partition_idx + 1, high)

File: Sort/QuickSort/QuickSort.java

/**
 * QuickSort implementation class.
 * @author Yasin Radi
 */
public class QuickSort {

    private static int partition(int[] array, int low, int high) {
        // Pivot
        int pivot = array[high];
        // Idx of smaller element
        int i = low - 1;
        int tmp = 0;

        for (int j = low; j < high; j++) {
            if (array[j] <= pivot) {
                i++;
                tmp = array[j];
                array[j] = array[i];
                array[i] = tmp;
            }
        }

        // Swap element with pivot
        tmp = array[i + 1];
        array[i + 1] = array[high];
        array[high] = tmp;
        
        return ++i;
    };

    /**
     * Sorts an array using QuickSort algorithm.
     * @param array int[]   Array to be sorted.
     * @param low   int     Lower array bound.
     * @param high  int     Higher array bound.
     */
    public static void sort(int[] array, int low, int high) {
        if (low < high) {
            int partition_idx = partition(array, low, high);

            // Sort first half
            sort(array, low, partition_idx - 1);

            // Sort second half
            sort(array, partition_idx + 1, high);
        }
    };
}